File: kotlin/src/test/kotlin/StatementPrinterTests.kt

import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Test
import org.approvaltests.Approvals.verify

class StatementPrinterTests {

    @Test
    internal fun exampleStatement() {

        val plays = mapOf(
            "hamlet" to Play("Hamlet", "tragedy"),
            "as-like" to Play("As You Like It", "comedy"),
            "othello" to Play("Othello", "tragedy")
        )

        val invoice = Invoice(
            "BigCo", listOf(
                Performance("hamlet", 55),
                Performance("as-like", 35),
                Performance("othello", 40)
            )
        )

        val statementPrinter = StatementPrinter()
        val result = statementPrinter.print(invoice, plays)

        verify(result)
    }

    @Test
    internal fun statementWithNewPlayTypes() {
        val plays = mapOf(
            "henry-v" to Play("Henry V", "history"),
            "as-like" to Play("As You Like It", "pastoral")
        )

        val invoice = Invoice(
            "BigCo", listOf(
                Performance("henry-v", 53),
                Performance("as-like", 55)
            )
        )

        val statementPrinter = StatementPrinter()
        assertThrows(Error::class.java) { statementPrinter.print(invoice, plays) }
    }
}

File: scala/src/test/scala/StatementPrinterTests.scala
import com.github.writethemfirst.approvals.Approvals
import com.github.writethemfirst.approvals.Approvals._
import org.scalatest.funsuite.AnyFunSuite
import org.scalatest.matchers.should.Matchers

class StatementPrinterTests extends AnyFunSuite with Matchers {
  test("statement example") {
    val plays = Map(
      "hamlet" -> Play("Hamlet", "tragedy"),
      "as-like" -> Play("As You Like It", "comedy"),
      "othello" -> Play("Othello", "tragedy")
    )

    val invoice = Invoice(
      "BigCo",
      List(
        Performance("hamlet", 55),
        Performance("as-like", 35),
        Performance("othello", 40)
      )
    )

    val statementPrinter = new StatementPrinter()
    val result = statementPrinter.print(invoice, plays)

    verify(result)
  }

  test("statement with new play types") {
    val plays = Map(
      "henry-v" -> Play("Henry V", "history"),
      "as-like" -> Play("As You Like It", "pastoral")
    )

    val invoice = Invoice(
      "BigCoII",
      List(
        Performance("henry-v", 53),
        Performance("as-like", 55)
      )
    )

    val statementPrinter = new StatementPrinter()
    assertThrows[Exception] {
      statementPrinter.print(invoice, plays)
    }
  }
}
